### PR TITLE
to_bytes in stdlib

### LINKED ISSUE #482: to_bytes in stdlib
# Problem

It's currently inconvenient to convert a `Field` element into `[u8]`. This seems like a common use case, e.g. when incorporating a `Field` element into a message for signature checks, or into the input of various hash functions.

# Solution

Define `to_bytes` as part of the stdlib.

# Alternatives considered

Document a convenient and recommended snippet for writing the above function directly into your circuit.

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves #482 

# Description

This PR adds a `to_bytes` to the Noir std lib to enable easy transform of a `Field` into `[u8]`. Having `to_bytes` natively in the compiler also uses many less gates in comparison to transforming a bit array to a byte array directly in Noir.

## Summary of changes

I added a `ToBytes` OPCODE as well as a new method `to_bytes` inside of `acir_gen.rs`. This acts in a very similar manner to the `split` method for bit decomposition except that we calculate the bytes values for decomposition through a nested for-loop where we split the bits by groups of 8, rather than one for loop where we calculate the bits like in `split`. 

I also added a new `Directive::ToBytes` for splitting bytes as the logic for filling in the witnesses differs from `Directive::Split`. I considered to change this to where I add another field to `Directive::Split` and based off this field determine whether to solve a bit or byte decomposition. I thought keeping the directives separate for now was the simpler solution.

## Dependency additions / changes

The `aztec_backend` had to be updated to reflect the new OPCODE. This PR can be found here: https://github.com/noir-lang/aztec_backend/pull/21

## Test additions / changes

I added a `to_bytes` test that transforms a field into a byte array.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

# Additional context

As per issue #548 we can rename the `Split` directive in a separate PR to follow the naming used here for byte decomposition.

### CODE REVIEW

**Comment by kevaundray** [comment]
Note: SplitBytes follows from Split, but its not very descriptive. In ACIR, we should have:

Split -> ToBits/BinaryDecomposition
SplitBytes -> ToBytes/ByteDecomposition

(The names on the right can be bikeshedded)

**Comment by vezenovm** [comment]
> Note: SplitBytes follows from Split, but its not very descriptive. In ACIR, we should have:
> 
> Split -> ToBits/BinaryDecomposition SplitBytes -> ToBytes/ByteDecomposition
> 
> (The names on the right can be bikeshedded)
> 
> --
> 
> We can rename the SplitBytes in this PR or atleast not follow the naming convention of split

I went with `Directive::ToBytes` for now. If we wish to go with the longer name we can, but I felt `ToBytes` was clear. We can then change split in a separate PR.

File: crates/acvm/src/lib.rs [hunk 1]
```diff
@@ -43,6 +43,7 @@ pub trait PartialWitnessGenerator {
         initial_witness: &mut BTreeMap<Witness, FieldElement>,
         gates: Vec<Gate>,
     ) -> GateResolution {
+        // dbg!(initial_witness.clone());
```
Comment (jfecher) [inline crates/acvm/src/lib.rs:46]: Can be removed

File: crates/noir_field/src/generic_ark.rs [hunk 2]
```diff
@@ -149,6 +149,10 @@ impl<F: PrimeField> FieldElement<F> {
         F::Params::MODULUS_BITS
     }
 
+    pub fn max_num_bytes() -> u32 {
+        Self::max_num_bits() / 8
```
Comment (jfecher) [inline crates/noir_field/src/generic_ark.rs:153]: Is max_num_bits guaranteed to be divisible by 8? If not we should add 1 if necessary

File: crates/noirc_evaluator/src/ssa/acir_gen.rs [hunk 3]
```diff
@@ -507,6 +507,9 @@ impl Acir {
                                         inputs.push(GadgetInput { witness: w, num_bits });
                                     }
                                 } else {
+                                    if array.values.is_empty() {
+                                        dbg!("inside prepare_inputs", array);
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:511]: dbg

File: crates/nargo/Cargo.toml [hunk 4]
```diff
@@ -28,7 +28,7 @@ hex = "0.4.2"
 tempdir = "0.3.7"
 
 # Backends
-aztec_backend = { optional = true, package = "barretenberg_static_lib", git = "https://github.com/noir-lang/aztec_backend", rev = "e1f206f6739d3782d3a241298089a0292de33742" }
+aztec_backend = { optional = true, package = "barretenberg_static_lib", git = "https://github.com/noir-lang/aztec_backend", rev = "ef6cdb1e26bff1c8dd9aea811e97b008f7e6651e" }
```
Comment (jfecher) [inline crates/nargo/Cargo.toml:31]: Why the change here?

Comment (vezenovm) [reply crates/nargo/Cargo.toml:31]: > The `aztec_backend` had to be updated to reflect the new OPCODE. This PR can be found here: [noir-lang/aztec_backend#21](https://github.com/noir-lang/aztec_backend/pull/21)

From the main PR description

File: crates/acvm/src/lib.rs (hunk 1 above)
Comment (vezenovm) [reply crates/acvm/src/lib.rs:46]: Done

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 3 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:511]: Done

File: crates/acir/src/circuit/gate.rs [hunk 5]
```diff
@@ -128,6 +129,15 @@ impl std::fmt::Debug for Gate {
                     b.last().unwrap().witness_index(),
                 )
             }
+            Gate::Directive(Directive::ToBytes { a, b, bit_size: _ }) => {
+                write!(
+                    f,
+                    "Split Bytes: {} into x{}...x{}",
```
Comment (kevaundray) [inline crates/acir/src/circuit/gate.rs:135]: Can you remove reference to Split Bytes here too?

File: crates/noirc_evaluator/src/ssa/acir_gen.rs [hunk 6]
```diff
@@ -630,6 +638,38 @@ pub fn evaluate_cmp(
...
+    let mut bytes = Expression::default();
+    let mut two_pow = FieldElement::one();
+    let two = FieldElement::from(2_i128);
+    let mut result = Vec::new();
+    for _ in 0..(bit_size / 8) {
+        let byte_witness = evaluator.add_witness_to_cs();
+        result.push(byte_witness);
+        let byte_expr = from_witness(byte_witness);
+        bytes = add(&bytes, two_pow, &byte_expr);
+        for _ in 0..8 {
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:653]: multiply by 256 instead

File: crates/noirc_evaluator/src/ssa/acir_gen.rs [hunk 7]
```diff
@@ -630,6 +638,38 @@ pub fn evaluate_cmp(
...
+            two_pow = two_pow.mul(two);
+        }
+        evaluator.gates.push(Gate::Arithmetic(subtract(
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:656]: why do you add this constraint which does not constrain anything?

File: crates/noirc_evaluator/src/ssa/acir_gen.rs [hunk 8]
```diff
@@ -630,6 +638,38 @@ pub fn evaluate_cmp(
...
+            &byte_expr,
+            FieldElement::one(),
+            &byte_expr,
+        )));
+    }
+    evaluator.gates.push(Gate::Directive(Directive::ToBytes {
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:662]: What about bit_size which are not a multiple of 8 or bigger than the modulus?

**Review by guipublic — CHANGES_REQUESTED** [review]
I think we should implement the to_radix function

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 7 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:656]: I actually wanted to ask about this. I basically wanted to constrain the byte expr I just created, however, I was wondering if I have to make it quadratic. Such as mul(byte_expr, byte_expr) - mul(byte_expr, byte_expr), or if this constraint is even required?

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 8 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:662]: For right now basically the last byte will be truncated if the bit_size is not a multiple of 8. And I have this assert statement, `assert!(bit_size < FieldElement::max_num_bits());`, at the top of `to_bytes`. It currently still follows what we are doing in the `split` function where the max bit size is 253 so it cannot be bigger than the modulus. 

I was thinking that checking directly against the modulus rather than the bit size might be better scoped in a separate PR.

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:662]: Thank you for your answer, I did not saw the assert, sorry.
Regarding the truncation if not multiple of 8, I think this is not good as it might fail and users will not know why, so you should either handle this case of force users to provide a multiple of 8 (i.e a byte_size rather than a bit_size)

File: crates/acir/src/opcode.rs [hunk 9]
```diff
@@ -35,6 +36,7 @@ impl OPCODE {
             OPCODE::EcdsaSecp256k1 => 7,
             OPCODE::FixedBaseScalarMul => 8,
             OPCODE::ToBits => 9,
+            OPCODE::ToBytes => 10,
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/acir/src/opcode.rs:39]: I went with `to_byte` to start as it was simpler, but I will look into how big of a change to_radix would be right now

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 8 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:661]: Ok. I am leaning towards requiring users to supply a byte size rather than a bit size. However, I am also trying to think of how to best handle this case for `to_radix` when we have different multiples, not just multiples of 8.

Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:661]: You could have the to_bytes in the stdlib, which requires a byte size, and this function uses to_radix internally.

File: crates/acir/src/opcode.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/acir/src/opcode.rs:39]: I think we should settle on exactly how we want to generalize `to_bits` and `to_bytes` as per my comment (https://github.com/noir-lang/noir/issues/482#issuecomment-1346993151) in the issue this PR aims to resolve. `to_bytes` provides useful functionality to the developer without much complexity on our end. I think we could look into adding the `to_bytes` function for now, and then generalizing `to_radix` in a separate PR.

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 7 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:660]: you want to constraint it to what?

File: crates/noir_field/src/generic_ark.rs [hunk 10]
```diff
@@ -149,6 +149,15 @@ impl<F: PrimeField> FieldElement<F> {
         F::Params::MODULUS_BITS
     }
 
+    pub fn max_num_bytes() -> u32 {
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/noir_field/src/generic_ark.rs:156]: Done

File: crates/acir/src/circuit/gate.rs (hunk 5 above)
Comment (vezenovm) [reply crates/acir/src/circuit/gate.rs]: Done

File: crates/acir/src/opcode.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/acir/src/opcode.rs:39]: The opcode should be to_radix, and we can expose to_bytes in the stdlib (at least for now)

[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/acir/src/opcode.rs:39]: Ok I will change this. We technically don't use either OPCODE in the backend too as we handle it in the noir compiler.

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 7 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:660]: I was originally thinking we had to constrain the intermediate decomposed expression, but I see how the constraint is kind of pointless now. 

Under the `split` function am I correct in assuming that this [line](https://github.com/noir-lang/noir/blob/dff8f393706c102055e78b4d97df207d73640eee/crates/noirc_evaluator/src/ssa/acir_gen.rs#L646) constrains the bit expression to be either a 1 or 0. I just tried to generalize `split` and `to_bytes` though with `to_radix` and I am now wondering if it is dangerous to remove this?

File: crates/acir/src/opcode.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/acir/src/opcode.rs:39]: I think we may be able to just remove these OPCODEs as ToBits and ToBytes are both builtin methods as per this issue #303

File: crates/acir/src/opcode.rs [hunk 11]
```diff
@@ -49,6 +51,7 @@ impl OPCODE {
             OPCODE::EcdsaSecp256k1 => "ecdsa_secp256k1",
             OPCODE::FixedBaseScalarMul => "fixed_base_scalar_mul",
             OPCODE::ToBits => "to_bits",
```
Comment (kevaundray) [inline crates/acir/src/opcode.rs:53]: Note to remove opcodes as we only need directives

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 7 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs]: I ended up using the `range_constraint` function already in the `acir_gen.rs` as I realized this is what I wanted. This is sufficient for constraining the generalized functionality of `split` and `to_bytes`.

File: crates/acir/src/opcode.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/acir/src/opcode.rs:39]: We can remove them in a separate PR

**Review by guipublic — APPROVED** [review]
That's fine for me.
We should do in another PR:
remove the opcodes and keep only the directive and builtin functions
use a to_radix method and directive
try to return unsized array in stdlib

**Review by jfecher — APPROVED** [review]
